geomesa-kafka-bundle/geomesa-kafka-processors/src/main/scala/org/geomesa/nifi/processors/kafka/GetGeoMesaKafkaRecord.scala

```diff
@@ -139,95 +134,111 @@ class GetGeoMesaKafkaRecord extends AbstractProcessor {
+            if (flowFile != null) {
+              Try(session.remove(flowFile)).failed.foreach(e.addSuppressed)
+            }
+            Failure(e)
         }
-      } catch {
-        case NonFatal(e) =>
-          logger.error("Error onTrigger:", e)
-          session.remove(flowFile)
       }
-    }
-  }
 
-  @tailrec
-  private def writeRemaining(writer: RecordSetWriter, timeout: Long, count: Int): Unit = {
-    if (timeout > System.currentTimeMillis()) {
-      val record = readNextRecord()
-      if (record != null) {
-        val count = writer.write(record).getRecordCount
-        if (count < maxBatchSize) {
-          writeRemaining(writer, timeout, count)
+      val errors = results.collect { case Failure(e) => e }
+      if (errors.nonEmpty) {
+        try {
+          val e = errors.head
+          errors.tail.foreach(e.addSuppressed)
+          logger.error("Error processing message batch:", e)
+          results.collect { case Success(f) => Try(session.remove(f)) }
+        } finally {
+          processor.done.put(false)
         }
-      } else if (count < minBatchSize) {
-        writeRemaining(writer, timeout, count)
+      } else {
+        var ok = true
+        results.foreach { case Success(f) =>
+          try {
+            if (f.getAttribute("record.count").toInt <= 0) {
+              logger.debug("Removing flow file, no records were written")
+              session.remove(f)
+            } else {
+              session.transfer(f, SuccessRelationship)
+            }
+          } catch {
+            case NonFatal(e) =>
+              ok = false
+              logger.error(s"Error transferring flow file $f:", e)
+              Try(session.remove(f))
+          }
+        }
+        processor.done.put(ok)
       }
     }
   }
 
-  private def readNextRecord(): Record = {
-    val feature = queue.poll(pollTimeout, TimeUnit.MILLISECONDS)
-    if (feature == null) { null } else {
-      converter.convert(feature)
-    }
-  }
-
   @OnRemoved
   @OnStopped
   @OnShutdown
   def cleanup(): Unit = {
     logger.info("Processor shutting down")
     val start = System.currentTimeMillis()
-    if (ds != null) {
-      fs.removeFeatureListener(listener)
-      CloseWithLogging(ds)
-      fs = null
-      ds = null
+    if (consumer != null) {
+      CloseWithLogging(consumer)
+      consumer = null
     }
     logger.info(s"Shut down in ${System.currentTimeMillis() - start}ms")
   }
 
-  class RecordFeatureListener extends FeatureListener {
-    override def changed(event: FeatureEvent): Unit = {
-      event match {
-        case e: KafkaFeatureChanged => queue.put(e.feature)
-        case _: KafkaFeatureRemoved => // no-op
-        case _: KafkaFeatureCleared => // no-op
-        case e                      => logger.error(s"Unknown event $e")
+  class RecordProcessor extends GeoMessageProcessor {
+
+    val ready = new SynchronousQueue[Seq[SimpleFeature]]()
+    val done = new SynchronousQueue[Boolean]()
```

Comment:
Due to the use of synchronous queues (and other sensible considerations), this processor should never have more than one thread.

Is it possible in the UI to add more threads?  

If it is presently possible, can we prevent that?

Reply:
it's definitely possible in the UI. i'm not sure of any way to prevent that, although maybe there are flags or annotations you can use. i'm also not 100% sure what happens when you use multiple threads in nifi (i.e. is onTrigger just called multiple times, or is a whole new processor spun up? at least in a clustered nifi, there would presumably be different instances of the processor on different machines). this probably requires more investigation, could maybe be a follow-up ticket.

Reply:
Do we at least have a note in the docs that this processor ought to be used with just one thread?

Reply:
yes it's here: https://github.com/locationtech/geomesa/blob/main/docs/user/nifi/ingress.rst#getgeomesakafkarecord